[DEVREL-29] allow pretty permalinks to be handled for rest endpoints by default

jazzsequence opened the pull request:
Fixes #178 

The linked issue is somewhat misrepresented. While the observed behavior is accurate (malformed URLs) it is _not_ because of a misplaced `/wp` in the REST url -- WP-API URLs on Bedrock-based sites are _expected_ to have two `/wp/` strings in them -- one being the path to the application and the second is part of the actual `wp-json/` URL (e.g. `/wp-json/wp/v2/...`). Therefore, the proposed solution would actually _break_ sites without addressing the issue.

Since the issue only exists on _newly created sites_, the assumption here is that these sites are _not using "pretty permalinks"_. Indeed, our tests _assume_ permalinks are pretty for this very reason. In cases when "pretty permalinks" are not active, it makes sense that API requests might fail due to our URL rewriting for internal paths.

This PR works around the issue by simply _accepting_ "pretty"-style API endpoint paths while also preserving the functionality of "plain" permalinks. Both "plain" and "pretty" permalinks are henceforth supported by the REST API for WPCM sites.

~Tests are broken out into #182 so as to not mix [release and non-release commits](https://github.com/pantheon-systems/wordpress-composer-managed/blob/default/README-internal.md#release-and-non-release-commits) and will fail until this PR is merged.~

Tests have been integrated back into this PR as well as other non-release changes. As a result, this PR MUST BE merged with a merge commit, **not** squash merge.

jazzsequence committed:
DEVREL-29: handle pretty permalinks for rest endpoints by default (#182)

bypass the issue with bad string replacement by just parsing pretty
permalinks natively for rest endpoints even when pretty permalinks
aren't active

jazzsequence committed:
allow pretty permalinks to be handled for rest endpoints by default
bypass the issue with bad string replacement by just parsing pretty permalinks natively for rest endpoints even when pretty permalinks aren't active

jazzsequence committed:
DEVREL-29: Update tests for handling plain permalinks (#182)

pwtyler commented:
> Tests are broken out into https://github.com/pantheon-systems/wordpress-composer-managed/pull/182 so as to not mix [release and non-release commits](https://github.com/pantheon-systems/wordpress-composer-managed/blob/default/README-internal.md#release-and-non-release-commits) and will fail until this PR is merged.

This is why "rebase and merge" exists. I'm going to cherry-pick them back over here.

pwtyler commented on web/app/mu-plugins/filters.php:
```diff
@@ -235,3 +235,149 @@ function __rebuild_url_from_parts( array $parts ) : string {
+	$rest_route = '/'; // Default to base route
+	$wp_json_pos = strpos( $path, '/wp-json/' );
+
+	if ( $wp_json_pos !== false ) {
+		$extracted_route = substr( $path, $wp_json_pos + strlen( '/wp-json' ) ); // Get everything after /wp-json
+		// Special case: Handle the originally reported '/wp-json/wp/' malformation
+		if ( strpos( $extracted_route, 'wp/' ) === 0 ) {
+			$extracted_route = substr( $extracted_route, strlen( 'wp' ) ); // Remove the extra 'wp'
+		}
+		// Ensure the extracted route starts with a slash
+		if ( ! $extracted_route && $extracted_route[0] !== '/' ) {
+			$extracted_route = '/' . $extracted_route;
+		}
+		$rest_route = $extracted_route ?: '/'; // Use extracted route or default to base
+	}
+	return $rest_route;
```
```suggestion
	if ( $wp_json_pos === false ) {
		return $rest_route;
	}
	$extracted_route = substr( $path, $wp_json_pos + strlen( '/wp-json' ) ); // Get everything after /wp-json
	// Special case: Handle the originally reported '/wp-json/wp/' malformation
	if ( strpos( $extracted_route, 'wp/' ) === 0 ) {
		$extracted_route = substr( $extracted_route, strlen( 'wp' ) ); // Remove the extra 'wp'
	}
	// Ensure the extracted route starts with a slash
	if ( ! $extracted_route && $extracted_route[0] !== '/' ) {
		$extracted_route = '/' . $extracted_route;
	}
	$rest_route = $extracted_route ?: '/'; // Use extracted route or default to base
```
disappointed. I thought you were on the return early train.

pwtyler commented on web/app/mu-plugins/filters.php:
```diff
@@ -235,3 +235,149 @@ function __rebuild_url_from_parts( array $parts ) : string {
+			// Directly set the query variable for the REST API.
+			$wp->query_vars['rest_route'] = $endpoint;
+
+			// Optional: Unset other query vars WP might have incorrectly parsed.
+			// unset($wp->query_vars['pagename'], $wp->query_vars['error']);
```
keeping this around? or delete
```suggestion
```

jazzsequence committed:
remove todo

Co-authored-by: Phil Tyler <philip@tylerdigital.com>

jazzsequence committed:
remove commented-out code

Co-authored-by: Phil Tyler <philip@tylerdigital.com>

jazzsequence committed:
Merge branch 'devrel-29-fix-subdir-wpms-url-issue-2' of github.com:pantheon-systems/wordpress-composer-managed into devrel-29-fix-subdir-wpms-url-issue-2

jazzsequence committed:
Merge branch 'default' into devrel-29-fix-subdir-wpms-url-issue-2

jazzsequence committed:
fix bats test

jazzsequence committed:
return early

jazzsequence committed:
remove redundant else

jazzsequence committed:
fix spacing

jazzsequence committed:
add helper functions

jazzsequence committed:
add teardown

jazzsequence committed:
use helper functions

jazzsequence committed:
add a test that checks the json body of the api endpoint for hello-world post

jazzsequence committed:
change delete site step to only run on success

jazzsequence committed:
remove the hard failure for mixed PRs

jazzsequence committed:
add a manual delete test sites workflow

jazzsequence committed:
fix path to bats test

jazzsequence committed:
we can't actually delete the subdomain test site
because we need a site plan to actually use subdomains

jazzsequence committed:
use bats-assert

jazzsequence committed:
log, but don't display errors

jazzsequence committed:
fix path to bats-assert

jazzsequence committed:
load bats-support

jazzsequence committed:
bats-assert and bats-support should be installed by the action

jazzsequence committed:
use bats-action standard way of loading libraries

jazzsequence committed:
remove export for BATS_LIB_PATH

jazzsequence committed:
don't use bats_load_library

jazzsequence committed:
set the lib path in the action env

jazzsequence committed:
move the BATS_LIB_PATH into the run command

jazzsequence committed:
try exporting locally again

jazzsequence committed:
filter out terminus noise from wp commands

jazzsequence committed:
use partial assertion for home url path

jazzsequence committed:
setup permalinks always

jazzsequence committed:
run a full env:clear-cache and wait before moving on

jazzsequence committed:
ensure permalinks are pretty for playwright tests
subdom installs don't run setup playwright tests script (unless the site doesn't exist) so we need to at least clean up after the bats tests maybe set permalinks to plain

jazzsequence committed:
fix terminus site name
copy pasta from setup script failed

jazzsequence committed:
fix workflow:wait

jazzsequence committed:
need site env and move wait

jazzsequence committed:
fix tests

jazzsequence committed:
standardize vars

jazzsequence committed:
don't require wp-cli commands for post id and url

jazzsequence committed:
just remove clear-cache

jazzsequence committed:
remove trap line to test

jazzsequence committed:
change assert_match to assert_success

jazzsequence committed:
click the continue button if the continue button exists

jazzsequence committed:
try a different means of clicking the button

jazzsequence committed:
adjust tests for more robust single/multisite handling

jazzsequence committed:
update service-level to plan

jazzsequence committed:
set the plan to a real plan

jazzsequence committed:
set the plan to free before delete

jazzsequence committed:
update changelog

jazzsequence committed:
dont run tests when delete test sites workflow is updated

jazzsequence committed:
downgrade sites before delete

jazzsequence committed:
bypass interstertial page via http headers

jazzsequence committed:
add || true to plan:set command
if it fails, it shouldn't prevent site deletion

jazzsequence committed:
test php 8.4 too

jazzsequence committed:
better handling of check commits step

jazzsequence committed:
style

jazzsequence committed:
ensure the comment only appears once

jazzsequence committed:
remove backticks
this shouldn't be a code comment

jazzsequence committed:
but maybe it should be a blockquote

jazzsequence committed:
combine multisite checks

jazzsequence commented:
Going to close this PR and open a new one with two clean commits.
